refactor: Mark CAddrMan::Select and GetAddr const

maflcko opened the pull request:
To clarify that a call to this only changes the random state and nothing else.

Commit:
refactor: Mark CAddrMan::Select const

Commit:
refactor: Mark CAddrMan::GetAddr const

promag reviewed (commented):
ACK fa8e3f78751d80fa8bce957387bfc95802af2477.

Built locally, also tried to avoid `mutable` members, but all things considered, this approach seems preferable.

jonatack commented on src/addrman.cpp:
```diff
@@ -394,8 +398,9 @@ CAddrInfo CAddrMan::Select_(bool newOnly)
             int nId = vvNew[nUBucket][nUBucketPos];
-            assert(mapInfo.count(nId) == 1);
-            CAddrInfo& info = mapInfo[nId];
+            const auto it_found{mapInfo.find(nId)};
```
Here and line 383 above, apparently using `auto` with `{}` can lead to surprising results and it's recommended to use `=` rather than `{}` for objects specified with `auto` when we don't mean a list (at least, for C++11 per my Stroustrup book, not sure if it changed since in 14 or 17, on first skim of https://en.cppreference.com/w/cpp/language/auto I didn't see a change).

MarcoFalke replied:
Mind to share an example to illustrate the "surprise"?

jonatack replied:
![image](https://user-images.githubusercontent.com/2415484/118260513-f3ec0e00-b4a1-11eb-95e7-8c1423580760.jpeg)

![image](https://user-images.githubusercontent.com/2415484/118260555-00706680-b4a2-11eb-8a95-6f8d85e52d9d.jpeg)

(afk, sent from phone)

MarcoFalke replied:
See https://en.cppreference.com/w/cpp/language/template_argument_deduction#auto_type_deduction

Also, if this was a list, the code wouldn't compile.

jonatack replied:
Thanks for the link. Seems it changed from 2014: https://isocpp.org/blog/2014/03/n3922, IIUC.

jonatack replied:
"For direct list-initialization:

    For a braced-init-list with only a single element, auto deduction will deduce from that entry;
    For a braced-init-list with more than one element, auto deduction will be ill-formed."

theStack reviewed (approved):
Code-review ACK fa8e3f78751d80fa8bce957387bfc95802af2477

MarcoFalke commented:
Rebased

theStack reviewed (approved):
re-ACK fa845812d94ec6adc13459dca708b0850a3059ca 🦉 
Checked that changes since my last ACK are only rebase-related via `git range-diff fa8e3f78...fa845812`

theStack reviewed (approved):
re-re-ACK fa49ab3e4075f8402e30507091033a621a6941ac ⚓ 

Checked that changes since my last re-ACK are only rebase-related (caused by 5cd7f8abe3996d303774b6cddeb419337d605d02 and 8caf60dbbe16afa3c52574a7f6710d74c0bfd4ab) via `git range-diff fa845812...fa49ab3`

MarcoFalke commented:
Rebased and added one commit

Commit:
fuzz: Actually use const addrman

lsilva01 reviewed (approved):
Code Review ACK and Tested ACK https://github.com/bitcoin/bitcoin/pull/21940/commits/fa47bdddbc7fc153c36f8f700099656ec1ce9d5d on Ubuntu 20.04.

theStack reviewed (approved):
re-ACK fa47bdddbc7fc153c36f8f700099656ec1ce9d5d

Checked again that changes since my last re-ACK are only rebase-related via `git range-diff fa49ab3...fa47bddd` (plus the fuzzing related extra commit)

jnewbery commented on src/addrman.h:
```diff
@@ -671,7 +671,7 @@ class CAddrMan
     std::unordered_map<CNetAddr, int, CNetAddrHash> mapAddr GUARDED_BY(cs);
 
     //! randomly-ordered vector of all nIds
```
Perhaps comment _why_ this can be mutable. `vRandom` is unobservable outside the addrman class, so any changes to it are also unobservable (and hence it can be mutated in const methods).

MarcoFalke replied:
Thanks, done

jnewbery reviewed (commented):
Concept ACK. `vRandom` and `nRandomPos` should be unobservable outside addrman. That means we can mutate them under const methods, and client code should be unaware of those changes.

Can `GetAddr()` also be made const?

MarcoFalke commented:
Force pushed to add comment

MarcoFalke commented:
> Can GetAddr() also be made const?

It already is. (Adjusted title)

sipa commented:
Concept ACK. Perhaps this deserves a comment in addrman.h though, to state that even `const` member functions cannot be used without synchronization?

MarcoFalke commented:
Added a lock annotation to the mutable member, which can be "read" by compilers and developers

Commit:
Add missing GUARDED_BY to CAddrMan::insecure_rand

Commit:
Fix incorrect whitespace in addrman

Leaving it as-is would be annoying because some editor fix-up the
spacing when opening a file or editing it.

jnewbery commented on src/addrman.h:
```diff
@@ -631,9 +631,8 @@ class CAddrMan
-    FastRandomContext insecure_rand;
+    mutable FastRandomContext insecure_rand GUARDED_BY(cs);
 
-private:
```
I don't really like changing all of addrman's private members to be protected. Since this is just to allow the tests to modify `nKey` and `insecure_rand`, what do you think about instead adding a `bool deterministic` argument to `CAddrMan`'s ctor (similar to the `deterministic` argument in `TxRequestTracker::TxRequestTracker`).

MarcoFalke replied:
Is there any risk allowing the unit tests access? Happy to add `private` back (after the Mutex) if you think it helps.

jnewbery replied:
Yeah, I'd prefer for `private` to be added after the mutex to minimize the change.

MarcoFalke replied:
Force pushed to add `private`

MarcoFalke commented on src/addrman.cpp:
```diff
@@ -150,11 +150,13 @@ void CAddrMan::SwapRandom(unsigned int nRndPos1, unsigned int nRndPos2)
-    assert(mapInfo.count(nId2) == 1);
+    const auto it_1{mapInfo.find(nId1)};
+    const auto it_2{mapInfo.find(nId2)};
+    assert(it_1 != mapInfo.end());
+    assert(it_2 != mapInfo.end());
 
-    mapInfo[nId1].nRandomPos = nRndPos2;
-    mapInfo[nId2].nRandomPos = nRndPos1;
+    it_1->second.nRandomPos = nRndPos2;
+    it_2->second.nRandomPos = nRndPos1;
```
I'll leave that behaviour change (Replacing `assert` with `throw`) for a follow-up

jnewbery replied:
It's not a behavior change - the code is already asserting that `count(nId1)` and `count(nId2)` are non-zero, done under the `cs` lock so the map can't change from under us.

MarcoFalke replied:
Obviously it doesn't matter here, but one lookup might be faster than two lookups, which is why I changed to `find`.

jnewbery replied:
Sure. Just a suggestion to minimize the diff (+2/-2 instead of +6/-4).

jnewbery commented:
Code review ACK fae108ceb53f61d7338ba205873623ede3c1d3be

MarcoFalke commented:
@promag or @theStack interested in donating a re-ACK?

theStack reviewed (approved):
re-ACK fae108ceb53f61d7338ba205873623ede3c1d3be 🍦 

Checked via `git range-diff fa47bddd...fae108ce` that there were only minor changes (comment on vRandom, as suggested by jnewbery) and rebase on master since my last ACK

jonatack commented:
Did of the reviewers test this change with DEBUG_ADDRMAN defined? It blows up for me.

jonatack commented:
Did of the reviewers test this change with DEBUG_ADDRMAN defined? It blows up for me.

```
addrman.cpp:467:15: error: out-of-line definition of 'Check_' does not match any declaration in 'CAddrMan'
int CAddrMan::Check_()
              ^~~~~~
./addrman.h:746:9: note: member declaration does not match because it is const qualified
    int Check_() const EXCLUSIVE_LOCKS_REQUIRED(cs);
        ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

MarcoFalke commented:
@jonatack Good catch. I didn't test this with the CPP flag set and CI doesn't either.

Luckily the code will be always compiled after #20233, so shouldn't happen again after that pull is merged.

jnewbery commented:
Thanks @jonatack. I'm currently rebasing #20233 and will include a fix for this issue in that PR.

jonatack commented:
> Thanks @jonatack. I'm currently rebasing #20233 and will include a fix for this issue in that PR.

Thanks!

MarcoFalke commented:
I created a fix in #22601 , because I didn't see the comment here.

jonatack commented:
Thanks, in the meantime will cherry-pick this commit locally.

theStack commented:
> Did any of the reviewers test this change with DEBUG_ADDRMAN defined? It blows up for me.

Oops, no, I didn't :( thanks for catching this.

Would it make sense to have a CI instance with DEBUG_ADDRMAN defined or is that overambitious? 🤔

jnewbery commented:
@theStack please see #20233 which removes the `DEBUG_ADDRMAN` build option and makes consistency checks a runtime option.
